python3Packages.swh-*: init

drupol opened the pull request:
This PR:

- Package some `swh-*` python modules: https://gitlab.softwareheritage.org/swh/devel/
- Package `attrs-strict`, required by `swh-model`: https://pypi.org/project/attrs-strict/

Status: all good !

![image](https://github.com/user-attachments/assets/a7c048b7-bae4-4117-b163-8c44781d3c6d)

Quickly test with: nix shell --impure --expr 'let pkgs = import (builtins.getFlake "github:drupol/nixpkgs/init/swh-model") { }; in pkgs.python3.withPackages (ps: with ps; [ swh-core swh-scanner ])'

Ping @zacchiro for review

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

drupol committed:
python312Packages.attrs-strict: init at 1.0.1

drupol committed:
python312Packages.swh-model: init at 7.1.0

drupol committed:
python312Packages.swh-core: init at 4.0.0

drupol committed:
python312Packages.swh-web-client: init at 0.9.0

drupol committed:
python312Packages.swh-auth: init at 0.10.0

mweinelt commented on pkgs/development/python-modules/attrs-strict/default.nix:
```diff
@@ -0,0 +1,38 @@
+}:
+
+buildPythonPackage rec {
+  pname = "attrs_strict";
```
```suggestion
  pname = "attrs-strict";
```

PEP503 normalization

mweinelt commented on pkgs/development/python-modules/attrs-strict/default.nix:
```diff
@@ -0,0 +1,38 @@
+  pname = "attrs_strict";
+  version = "1.0.1";
+  disabled = pythonOlder "3.7";
+  format = "pyproject";
```
> 
```suggestion
  pyproject = true;
```

mweinelt commented on pkgs/development/python-modules/attrs-strict/default.nix:
```diff
@@ -0,0 +1,38 @@
+    hash = "sha256-5wSGNiAUbF8qi2Ac1FdNFIkT2yb8Bjb5Qf5CEuQl6v4=";
+  };
+
+  build-system = [
+    python3.pkgs.setuptools
+    python3.pkgs.setuptools-scm
+  ];
+
+  dependencies = with python3.pkgs; [
+    attrs
+  ];
```
Consume these directly as function args when packaging within python-modules.

The way you are currently packaging them makes them stick to the current `python3` and prevent multiple versioned package sets.

mweinelt commented on pkgs/development/python-modules/attrs-strict/default.nix:
```diff
@@ -0,0 +1,38 @@
+    homepage = "https://github.com/bloomberg/attrs-strict";
+    changelog = "https://github.com/bloomberg/attrs-strict/releases/tag/${version}";
+    license = lib.licenses.asl20;
+    maintainers = with lib.maintainers; [ ];
```
If you want to have them, you'll have to maintain them.

drupol replied:
Fixed

mweinelt reviewed (dismissed):
A bunch of generic issues that must be fixed across all packages.

drupol committed:
python312Packages.swh-scanner: init at 0.8.3

drupol committed:
python312Packages.mirakuru: unbreak build by disabling one test

JulienMalka reviewed (approved):
This looks good to me. All items from preceding reviews seem fixed so I'll merge this later today if there is no more feedback.

drupol committed:
swh: init at 4.0.0

Wrap `swh` application with `writeShellApplication` to avoid exposing internal Python modules.

Using `writeShellApplication` ensures that the underlying Python packages and modules remain hidden.  
`toPythonApplication` could not be used due to infinite recursion issues caused by circular dependencies between Python modules.

drupol commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 315400`

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 25 packages built:</summary>
  <ul>
    <li>python312Packages.attrs-strict</li>
    <li>python312Packages.attrs-strict.dist</li>
    <li>python312Packages.swh-auth</li>
    <li>python312Packages.swh-auth.dist</li>
    <li>python312Packages.swh-core</li>
    <li>python312Packages.swh-core.dist</li>
    <li>python312Packages.swh-model</li>
    <li>python312Packages.swh-model.dist</li>
    <li>python312Packages.swh-scanner</li>
    <li>python312Packages.swh-scanner.dist</li>
    <li>python312Packages.swh-web-client</li>
    <li>python312Packages.swh-web-client.dist</li>
    <li>python313Packages.attrs-strict</li>
    <li>python313Packages.attrs-strict.dist</li>
    <li>python313Packages.swh-auth</li>
    <li>python313Packages.swh-auth.dist</li>
    <li>python313Packages.swh-core</li>
    <li>python313Packages.swh-core.dist</li>
    <li>python313Packages.swh-model</li>
    <li>python313Packages.swh-model.dist</li>
    <li>python313Packages.swh-scanner</li>
    <li>python313Packages.swh-scanner.dist</li>
    <li>python313Packages.swh-web-client</li>
    <li>python313Packages.swh-web-client.dist</li>
    <li>swh</li>
  </ul>
</details>

drupol commented:
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 315400`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 33 packages built:</summary>
  <ul>
    <li>python312Packages.attrs-strict</li>
    <li>python312Packages.attrs-strict.dist</li>
    <li>python312Packages.mirakuru</li>
    <li>python312Packages.mirakuru.dist</li>
    <li>python312Packages.pytest-postgresql</li>
    <li>python312Packages.pytest-postgresql.dist</li>
    <li>python312Packages.swh-auth</li>
    <li>python312Packages.swh-auth.dist</li>
    <li>python312Packages.swh-core</li>
    <li>python312Packages.swh-core.dist</li>
    <li>python312Packages.swh-model</li>
    <li>python312Packages.swh-model.dist</li>
    <li>python312Packages.swh-scanner</li>
    <li>python312Packages.swh-scanner.dist</li>
    <li>python312Packages.swh-web-client</li>
    <li>python312Packages.swh-web-client.dist</li>
    <li>python313Packages.attrs-strict</li>
    <li>python313Packages.attrs-strict.dist</li>
    <li>python313Packages.mirakuru</li>
    <li>python313Packages.mirakuru.dist</li>
    <li>python313Packages.pytest-postgresql</li>
    <li>python313Packages.pytest-postgresql.dist</li>
    <li>python313Packages.swh-auth</li>
    <li>python313Packages.swh-auth.dist</li>
    <li>python313Packages.swh-core</li>
    <li>python313Packages.swh-core.dist</li>
    <li>python313Packages.swh-model</li>
    <li>python313Packages.swh-model.dist</li>
    <li>python313Packages.swh-scanner</li>
    <li>python313Packages.swh-scanner.dist</li>
    <li>python313Packages.swh-web-client</li>
    <li>python313Packages.swh-web-client.dist</li>
    <li>swh</li>
  </ul>
</details>

mweinelt commented on pkgs/development/python-modules/swh-core/default.nix:
```diff
@@ -0,0 +1,127 @@
+    "swh/core/api/tests/test_rpc_server_asynchronous.py"
+    # ModuleNotFoundError: No module named 'systemd'
+    "swh/core/tests/test_logger.py"
+    # ModuleNotFoundError: No module named 'psycopg_pool'
+    "swh/core/db/tests"
```
Available as `psycopg.optional-dependencies.pool`

mweinelt commented on pkgs/development/python-modules/swh-core/default.nix:
```diff
@@ -0,0 +1,127 @@
+    # ModuleNotFoundError: No module named 'aiohttp_utils'
+    "swh/core/api/tests/test_async.py"
+    "swh/core/api/tests/test_rpc_server_asynchronous.py"
+    # ModuleNotFoundError: No module named 'systemd'
+    "swh/core/tests/test_logger.py"
```
Available as `systemd`, but I'm sure you found that out already.

mweinelt commented on pkgs/development/python-modules/attrs-strict/default.nix:
```diff
@@ -0,0 +1,42 @@
+
+  pythonImportsCheck = [ "attrs_strict" ];
+
+  # No tests in the pypi archive
+  doCheck = false;
```
The upstream repo has tags, so fetch that instead.
